File: src/lib/services/session.ts
import { SupabaseClient } from '@supabase/supabase-js';
import { DOOR_OPEN_FRAUD_SECONDS } from '../config';
import type { DbSession, DbTariff } from '../types';

export class SessionService {
  constructor(private db: SupabaseClient) {}

  async startSession(userId: string, cellId: string, bookingId?: string | null): Promise<DbSession> {
    const { data: user } = await this.db.from('users').select('*').eq('id', userId).maybeSingle();
    if (!user) throw new Error('User not found');
    if (user.has_debt) throw new Error(`User has debt: ${user.debt_amount}`);

    const { count: activeSessions } = await this.db
      .from('sessions')
      .select('*', { count: 'exact', head: true })
      .eq('user_id', userId)
      .is('end_at', null);
    if ((activeSessions ?? 0) >= 2) throw new Error('Max active sessions reached');

    const { data: cell } = await this.db.from('cells').select('id').eq('id', cellId).maybeSingle();
    if (!cell) throw new Error('Cell not found');

    if (bookingId) {
      const { data: booking } = await this.db
        .from('bookings')
        .select('status')
        .eq('id', bookingId)
        .maybeSingle();
      if (booking && booking.status === 'ACTIVE') {
        await this.db.from('bookings').update({ status: 'CONVERTED' }).eq('id', bookingId);
      }
    }

    const now = new Date();
    const { data: session, error } = await this.db
      .from('sessions')
      .insert({
        user_id: userId,
        cell_id: cellId,
        booking_id: bookingId || null,
        start_at: now.toISOString(),
      })
      .select()
      .single<DbSession>();
    if (error) throw new Error(error.message);

    await this.db.from('cells').update({ status: 'BUSY' }).eq('id', cellId);
    return session;
  }

  async endSession(
    sessionId: string,
    doorClosed: boolean,
    chargerDisconnected: boolean,
  ): Promise<DbSession> {
    if (!doorClosed) throw new Error('Door must be closed before ending session');
    if (!chargerDisconnected) throw new Error('Charger must be disconnected before ending session');

    const { data: session } = await this.db
      .from('sessions')
      .select('*')
      .eq('id', sessionId)
      .maybeSingle<DbSession>();
    if (!session) throw new Error('Session not found');
    if (session.end_at !== null) throw new Error('Session already ended');

    const now = new Date();
    const startAt = new Date(session.start_at);
    const durationMins = (now.getTime() - startAt.getTime()) / 60000;

    const tariff = await this.getTariff(session.user_id, now);
    const cost = this.calculateCost(startAt, now, tariff);

    const { data: ended, error } = await this.db
      .from('sessions')
      .update({ end_at: now.toISOString(), duration_mins: durationMins, cost })
      .eq('id', sessionId)
      .select()
      .single<DbSession>();
    if (error) throw new Error(error.message);

    await this.db.from('cells').update({ status: 'FREE' }).eq('id', session.cell_id);

    if (cost > 0) {
      const { PaymentService } = await import('./payment');
      const paymentSvc = new PaymentService(this.db);
      await paymentSvc.charge(session.user_id, cost, sessionId);
    }

    return ended;
  }

  async getTariff(userId: string, now: Date): Promise<DbTariff | null> {
    const { data: sub } = await this.db
      .from('subscriptions')
      .select('tariff_id')
      .eq('user_id', userId)
      .eq('is_active', true)
      .gt('end_at', now.toISOString())
      .maybeSingle();

    if (sub) {
      const { data: tariff } = await this.db
        .from('tariffs')
        .select('*')
        .eq('id', sub.tariff_id)
        .maybeSingle<DbTariff>();
      if (tariff) return tariff;
    }

    const hour = now.getHours();
    const isNight = hour >= 22 || hour < 6;
    const { data: tariff } = await this.db
      .from('tariffs')
      .select('*')
      .eq('is_subscription', false)
      .eq('is_night', isNight)
      .maybeSingle<DbTariff>();
    if (tariff) return tariff;

    const { data: fallback } = await this.db
      .from('tariffs')
      .select('*')
      .eq('is_subscription', false)
      .maybeSingle<DbTariff>();
    return fallback;
  }

  calculateCost(start: Date, end: Date, tariff: DbTariff | null): number {
    if (!tariff) return 0;
    const durationMins = (end.getTime() - start.getTime()) / 60000;
    const freeMins = tariff.free_mins ?? 0;
    const billableMins = Math.max(0, durationMins - freeMins);
    let cost = billableMins * tariff.price_per_minute;
    if (tariff.discount_pct && tariff.discount_pct > 0) {
      cost = cost * (1 - tariff.discount_pct / 100);
    }
    return Math.round(cost * 100) / 100;
  }

  async forceCancelSession(sessionId: string): Promise<void> {
    const { data: session } = await this.db
      .from('sessions')
      .select('*')
      .eq('id', sessionId)
      .maybeSingle<DbSession>();
    if (!session || session.end_at !== null) return;

    const now = new Date();
    const startAt = new Date(session.start_at);
    const elapsed = (now.getTime() - startAt.getTime()) / 1000;
    if (elapsed < DOOR_OPEN_FRAUD_SECONDS) return;

    const durationMins = elapsed / 60;
    const tariff = await this.getTariff(session.user_id, now);
    const cost = this.calculateCost(startAt, now, tariff);

    await this.db
      .from('sessions')
      .update({ end_at: now.toISOString(), duration_mins: durationMins, cost })
      .eq('id', sessionId);
    await this.db.from('cells').update({ status: 'FREE' }).eq('id', session.cell_id);

    if (cost > 0) {
      const { PaymentService } = await import('./payment');
      const paymentSvc = new PaymentService(this.db);
      await paymentSvc.charge(session.user_id, cost, sessionId);
    }
  }
}

File: backend/app/services/session.py
from datetime import datetime, timedelta, timezone
from app.config import DOOR_OPEN_FRAUD_SECONDS
import logging

logger = logging.getLogger(__name__)


class SessionService:
    def __init__(self, db):
        self.db = db

    async def start_session(self, user_id: str, cell_id: str, booking_id: str | None = None):
        user = await self.db.user.find_unique(where={"id": user_id})
        if not user:
            raise ValueError("User not found")
        if user.hasDebt:
            raise ValueError(f"User has debt: {user.debtAmount}")

        active_sessions = await self.db.session.count(
            where={"userId": user_id, "endAt": None}
        )
        if active_sessions >= 2:
            raise ValueError("Max active sessions reached")

        cell = await self.db.cell.find_unique(where={"id": cell_id})
        if not cell:
            raise ValueError("Cell not found")

        if booking_id:
            booking = await self.db.booking.find_unique(where={"id": booking_id})
            if booking and booking.status == "ACTIVE":
                await self.db.booking.update(
                    where={"id": booking_id},
                    data={"status": "CONVERTED"},
                )

        now = datetime.now(timezone.utc)
        session = await self.db.session.create(data={
            "userId": user_id,
            "cellId": cell_id,
            "bookingId": booking_id,
            "startAt": now,
        })
        await self.db.cell.update(where={"id": cell_id}, data={"status": "BUSY"})
        return session

    async def end_session(self, session_id: str, door_closed: bool, charger_disconnected: bool):
        if not door_closed:
            raise ValueError("Door must be closed before ending session")
        if not charger_disconnected:
            raise ValueError("Charger must be disconnected before ending session")

        session = await self.db.session.find_unique(
            where={"id": session_id},
            include={"cell": {"include": {"locker": True}}},
        )
        if not session:
            raise ValueError("Session not found")
        if session.endAt is not None:
            raise ValueError("Session already ended")

        now = datetime.now(timezone.utc)
        duration_mins = (now - session.startAt).total_seconds() / 60

        # Get applicable tariff
        tariff = await self._get_tariff(session.userId, now)
        cost = self.calculate_cost(session.startAt, now, tariff)

        ended = await self.db.session.update(
            where={"id": session_id},
            data={
                "endAt": now,
                "durationMins": duration_mins,
                "cost": cost,
            },
        )

        await self.db.cell.update(where={"id": session.cellId}, data={"status": "FREE"})

        # Process payment
        if cost > 0:
            from app.services.payment import PaymentService
            payment_svc = PaymentService(self.db)
            await payment_svc.charge(session.userId, cost, session_id)

        return ended

    async def _get_tariff(self, user_id: str, now: datetime):
        sub = await self.db.subscription.find_first(
            where={"userId": user_id, "isActive": True, "endAt": {"gt": now}}
        )
        if sub:
            tariff = await self.db.tariff.find_unique(where={"id": sub.tariffId})
            if tariff:
                return tariff

        hour = now.hour
        is_night = hour >= 22 or hour < 6
        tariff = await self.db.tariff.find_first(
            where={"isSubscription": False, "isNight": is_night}
        )
        if not tariff:
            tariff = await self.db.tariff.find_first(where={"isSubscription": False})
        return tariff

    def calculate_cost(self, start: datetime, end: datetime, tariff) -> float:
        if tariff is None:
            return 0.0
        duration_mins = (end - start).total_seconds() / 60
        free_mins = tariff.freeMins if tariff.freeMins else 0
        billable_mins = max(0, duration_mins - free_mins)
        cost = billable_mins * tariff.pricePerMinute
        if tariff.discountPct and tariff.discountPct > 0:
            cost = cost * (1 - tariff.discountPct / 100)
        return round(cost, 2)

    async def force_cancel_session(self, session_id: str):
        """Anti-fraud: door held open too long."""
        session = await self.db.session.find_unique(where={"id": session_id})
        if not session or session.endAt is not None:
            return

        now = datetime.now(timezone.utc)
        elapsed = (now - session.startAt).total_seconds()
        if elapsed < DOOR_OPEN_FRAUD_SECONDS:
            return

        duration_mins = elapsed / 60
        tariff = await self._get_tariff(session.userId, now)
        cost = self.calculate_cost(session.startAt, now, tariff)

        await self.db.session.update(
            where={"id": session_id},
            data={"endAt": now, "durationMins": duration_mins, "cost": cost},
        )
        await self.db.cell.update(where={"id": session.cellId}, data={"status": "FREE"})

        if cost > 0:
            from app.services.payment import PaymentService
            payment_svc = PaymentService(self.db)
            await payment_svc.charge(session.userId, cost, session_id)

        logger.warning(f"Force-cancelled session {session_id} after {elapsed:.0f}s")
